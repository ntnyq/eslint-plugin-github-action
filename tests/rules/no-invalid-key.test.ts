import { expect } from 'vitest'
import rule, { RULE_NAME } from '../../src/rules/no-invalid-key'
import { $, run } from '../internal'

run({
  name: RULE_NAME,
  rule,
  valid: [
    {
      filename: 'top-level-keys.yml',
      code: $`
        name: CI
        
        run-name: CI
        
        on: push
        
        permissions:
          contents: read
          pull-requests: write
        
        env:
          SERVER: production
        
        defaults:
          run:
            shell: bash
        
        concurrency:
          group: example
          cancel-in-progress: true
        
        jobs:
          test:
            name: Test
            runs-on: ubuntu-latest
            steps:
              - name: Checkout
                uses: actions/checkout@v4
              - name: Run tests
                run: npm test
      `,
    },
    {
      filename: 'job-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            name: Unit Test
        
            permissions:
              contents: read
        
            needs: [lint]
        
            if: github.repository == 'ntnyq/eslint-plugin-github-action'
        
            runs-on: ubuntu-latest
        
            environment:
              name: Test
        
            concurrency:
              group: test
              cancel-in-progress: true
        
            outputs:
              test: true
        
            env:
              SERVER: production
        
            defaults:
              run:
                shell: bash
        
            steps:
              - run: npm run test
      `,
    },
    {
      filename: 'step-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            steps:
              - id: test
                if: github.repository == 'ntnyq/eslint-plugin-github-action'
                name: Checkout
                uses: actions/checkout@v4
                run: npm run test
                working-directory: ./src
                shell: bash
                with:
                  fetch-depth: 1
                env:
                  SERVER: production
                continue-on-error: true
                timeout-minutes: 10
      `,
    },
    {
      filename: 'strategy-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            strategy:
              fail-fast: true
              matrix:
                node-version: [14.x, 16.x]
                os: [ubuntu-latest, windows-latest, macos-latest]
              max-parallel: 3
      `,
    },
    {
      filename: 'container-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            container:
              image: ghcr.io/owner/image
              credentials:
                name: ntnyq
              env:
                SERVER: production
              ports:
                - 8080:8080
              volumes:
                - /var/run/docker.sock:/var/run/docker.sock
              options: --privileged
      `,
    },
    {
      filename: 'service-keys.yml',
      code: $`
        name: CI
        
        jobs:
          serve:
            services:
              nginx:
                image: ghcr.io/owner/image
                credentials:
                  name: ntnyq
                env:
                  SERVER: production
                ports:
                  - 8080:8080
                volumes:
                  - /var/run/docker.sock:/var/run/docker.sock
                options: --privileged
      `,
    },
  ],
  invalid: [
    {
      filename: 'top-level-keys.yml',
      code: $`
        workflow: CI
        
        dispatch: inputs
        
        push: branch
        
        test:
          contents: read
        
        check:
          SERVER: production
        
        matrix: [20.x, 22.x]
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      filename: 'job-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            workflow: CI
        
            dispatch: inputs
        
            push: branch
        
            test:
              contents: read
        
            check:
              SERVER: production
        
            matrix: [20.x, 22.x]
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      filename: 'step-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            steps:
              - workflow: CI
        
                dispatch: inputs
        
                push: branch
        
                test:
                  contents: read
        
                check:
                  SERVER: production
        
                matrix: [20.x, 22.x]
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      filename: 'strategy-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            strategy:
              workflow: CI
        
              dispatch: inputs
        
              push: branch
        
              test:
                contents: read
        
              check:
                SERVER: production
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      filename: 'container-keys.yml',
      code: $`
        name: CI
        
        jobs:
          test:
            container:
              workflow: CI
        
              dispatch: inputs
        
              push: branch
        
              test:
                contents: read
        
              check:
                SERVER: production
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
    {
      filename: 'service-keys.yml',
      code: $`
        name: CI
        
        jobs:
          serve:
            services:
              nginx:
                workflow: CI
        
                dispatch: inputs
        
                push: branch
        
                test:
                  contents: read
        
                check:
                  SERVER: production
      `,
      errors(errors) {
        expect(errors).toMatchSnapshot()
      },
    },
  ],
})
