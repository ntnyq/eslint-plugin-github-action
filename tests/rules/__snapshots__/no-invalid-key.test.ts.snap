// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-invalid-key > invalid > Invalid #0: workflow: CI

dispatch: inputs

push: branch

test:
  contents: read

check:
  SERVER: production

matrix: [20.x, 22.x] 1`] = `
[
  {
    "column": 1,
    "endColumn": 9,
    "endLine": 1,
    "line": 1,
    "message": "invalid top-level key \`workflow\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 9,
    "endLine": 3,
    "line": 3,
    "message": "invalid top-level key \`dispatch\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 5,
    "endLine": 5,
    "line": 5,
    "message": "invalid top-level key \`push\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 5,
    "endLine": 7,
    "line": 7,
    "message": "invalid top-level key \`test\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 6,
    "endLine": 10,
    "line": 10,
    "message": "invalid top-level key \`check\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 13,
    "line": 13,
    "message": "invalid top-level key \`matrix\`.",
    "messageId": "invalidTopLevelKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;

exports[`no-invalid-key > invalid > Invalid #1: name: CI

jobs:
  test:
    workflow: CI

    dispatch: inputs

    push: branch

    test:
      contents: read

    check:
      SERVER: production

    matrix: [20.x, 22.x] 1`] = `
[
  {
    "column": 5,
    "endColumn": 13,
    "endLine": 5,
    "line": 5,
    "message": "invalid job key \`workflow\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 5,
    "endColumn": 13,
    "endLine": 7,
    "line": 7,
    "message": "invalid job key \`dispatch\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 5,
    "endColumn": 9,
    "endLine": 9,
    "line": 9,
    "message": "invalid job key \`push\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 5,
    "endColumn": 9,
    "endLine": 11,
    "line": 11,
    "message": "invalid job key \`test\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 5,
    "endColumn": 10,
    "endLine": 14,
    "line": 14,
    "message": "invalid job key \`check\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 5,
    "endColumn": 11,
    "endLine": 17,
    "line": 17,
    "message": "invalid job key \`matrix\`.",
    "messageId": "invalidJobKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;

exports[`no-invalid-key > invalid > Invalid #2: name: CI

jobs:
  test:
    steps:
      - workflow: CI

        dispatch: inputs

        push: branch

        test:
          contents: read

        check:
          SERVER: production

        matrix: [20.x, 22.x] 1`] = `
[
  {
    "column": 9,
    "endColumn": 17,
    "endLine": 6,
    "line": 6,
    "message": "invalid step key \`workflow\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 17,
    "endLine": 8,
    "line": 8,
    "message": "invalid step key \`dispatch\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 13,
    "endLine": 10,
    "line": 10,
    "message": "invalid step key \`push\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 13,
    "endLine": 12,
    "line": 12,
    "message": "invalid step key \`test\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 14,
    "endLine": 15,
    "line": 15,
    "message": "invalid step key \`check\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 15,
    "endLine": 18,
    "line": 18,
    "message": "invalid step key \`matrix\`.",
    "messageId": "invalidStepKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;

exports[`no-invalid-key > invalid > Invalid #3: name: CI

jobs:
  test:
    strategy:
      workflow: CI

      dispatch: inputs

      push: branch

      test:
        contents: read

      check:
        SERVER: production 1`] = `
[
  {
    "column": 7,
    "endColumn": 15,
    "endLine": 6,
    "line": 6,
    "message": "invalid strategy key \`workflow\`.",
    "messageId": "invalidStrategyKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 15,
    "endLine": 8,
    "line": 8,
    "message": "invalid strategy key \`dispatch\`.",
    "messageId": "invalidStrategyKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 11,
    "endLine": 10,
    "line": 10,
    "message": "invalid strategy key \`push\`.",
    "messageId": "invalidStrategyKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 11,
    "endLine": 12,
    "line": 12,
    "message": "invalid strategy key \`test\`.",
    "messageId": "invalidStrategyKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 12,
    "endLine": 15,
    "line": 15,
    "message": "invalid strategy key \`check\`.",
    "messageId": "invalidStrategyKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;

exports[`no-invalid-key > invalid > Invalid #4: name: CI

jobs:
  test:
    container:
      workflow: CI

      dispatch: inputs

      push: branch

      test:
        contents: read

      check:
        SERVER: production 1`] = `
[
  {
    "column": 7,
    "endColumn": 15,
    "endLine": 6,
    "line": 6,
    "message": "invalid container key \`workflow\`.",
    "messageId": "invalidContainerKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 15,
    "endLine": 8,
    "line": 8,
    "message": "invalid container key \`dispatch\`.",
    "messageId": "invalidContainerKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 11,
    "endLine": 10,
    "line": 10,
    "message": "invalid container key \`push\`.",
    "messageId": "invalidContainerKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 11,
    "endLine": 12,
    "line": 12,
    "message": "invalid container key \`test\`.",
    "messageId": "invalidContainerKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 7,
    "endColumn": 12,
    "endLine": 15,
    "line": 15,
    "message": "invalid container key \`check\`.",
    "messageId": "invalidContainerKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;

exports[`no-invalid-key > invalid > Invalid #5: name: CI

jobs:
  serve:
    services:
      nginx:
        workflow: CI

        dispatch: inputs

        push: branch

        test:
          contents: read

        check:
          SERVER: production 1`] = `
[
  {
    "column": 9,
    "endColumn": 17,
    "endLine": 7,
    "line": 7,
    "message": "invalid service key \`workflow\`.",
    "messageId": "invalidServiceKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 17,
    "endLine": 9,
    "line": 9,
    "message": "invalid service key \`dispatch\`.",
    "messageId": "invalidServiceKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 13,
    "endLine": 11,
    "line": 11,
    "message": "invalid service key \`push\`.",
    "messageId": "invalidServiceKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 13,
    "endLine": 13,
    "line": 13,
    "message": "invalid service key \`test\`.",
    "messageId": "invalidServiceKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
  {
    "column": 9,
    "endColumn": 14,
    "endLine": 16,
    "line": 16,
    "message": "invalid service key \`check\`.",
    "messageId": "invalidServiceKey",
    "nodeType": "YAMLMapping",
    "ruleId": "no-invalid-key",
    "severity": 2,
  },
]
`;
